[wallet-ext] - account locking / unlocking + mnemonic import

mamos-mysten opened the pull request:
## Description 

Still a few things to work out, but this adds a generic way to dispatch the password unlock modal for accounts that are password unlockable. Also includes the ability to import a passphrase from the onboarding screen.


https://github.com/MystenLabs/sui/assets/122397493/77162446-90a3-42ec-8bf5-7c64db48aad5


## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+
+const UnlockAccountContext = createContext<UnlockAccountContextType | undefined>(undefined);
+
+export const UnlockAccountProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
```
super: It doesn't matter to me how we define/type components, but it would be nice to be consistent. For example, this versus the other provider further above:

`export const OnboardingFormProvider = ({ children }: { children: ReactNode }) =>`

mamos-mysten replied:
not even sure where that came from!

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+export const UnlockAccountProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
+	const [isUnlockModalOpen, setIsUnlockModalOpen] = useState(false);
+	const [accountIdToUnlock, setAccountIdToUnlock] = useState<string | null>(null);
+	const { data: accounts } = useAccounts();
```
I had a similar comment above, but can we use `useActiveAccount`?

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+					try {
+						// try unlocking non-password unlockable accounts immediately
+						await unlockAccountMutation.mutate({ id: accountId });
+						toast.success('Account unlocked');
```
I found it a little strange in the screen recording where you click unlock -> the loading spinner starts -> the toast appears saying the account was unlocked -> the loading spinner disappears shortly after. It feels like the loading spinner should stop first then the toast should appear right?

I think fixing the issue from above will fix this 🤔

mamos-mysten replied:
fixed

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+				} else {
+					try {
+						// try unlocking non-password unlockable accounts immediately
+						await unlockAccountMutation.mutate({ id: accountId });
```
nit: I'd do the success and error handling in the mutation callbacks instead of this try catch block (I also thought `mutate` wasn't async). 

https://tkdodo.eu/blog/mastering-mutations-in-react-query#mutate-or-mutateasync

```
unlockAccountMutation.mutate({ id: accountId }, {
  onSuccess: () => toast.success('Account unlocked'),
  onError: (e) => toast.error((e as Error).message),
});
```

pchrysochoidis replied:
yeah this should change to mutateAsync

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+
+	const lockAccount = useCallback(
+		async (accountId: string) => {
+			try {
```
same comment regarding the use of `mutate`

williamrobertson13 commented on apps/wallet/src/ui/app/components/menu/content/AccountActions.tsx:
```diff
@@ -20,16 +20,17 @@ export type AccountActionsProps = {
 
 export function AccountActions({ account }: AccountActionsProps) {
-	const exportAccountUrl = useNextMenuUrl(true, `/export/${account.address}`);
+	const exportAccountUrl = useNextMenuUrl(true, `/export/${account?.address}`);
```
isn't `account` always defined here?

williamrobertson13 commented on apps/wallet/src/ui/app/pages/accounts/manage/AccountGroup.tsx:
```diff
@@ -66,14 +69,24 @@ export function AccountGroup({
+								try {
+									// todo: support "Add New" for more account types
+									if (type === 'mnemonic') {
+										createAccountMutation.mutate({
```
`mutate` isn't `async` so same advice for using the `onSuccess` and `onError` callbacks

williamrobertson13 commented on apps/wallet/src/ui/app/pages/accounts/manage/ManageAccountsPage.tsx:
```diff
@@ -1,20 +1,45 @@
+				{Object.entries(groupedAccounts).map(([type, accounts]) => (
+					<AccountGroup
+						key={type}
+						accountSource={''}
```
is this empty because ZK accounts don't have an account source? could we add a fixme / todo if this needs re-worked?

mamos-mysten replied:
yea, this needs a `todo:`

williamrobertson13 commented on apps/wallet/src/ui/app/pages/layout/index.tsx:
```diff
@@ -23,17 +24,19 @@ function PageLayout({ forceFullscreen = false, children, className }: PageLayout
-				<div id="toaster-portal-container"></div>
-			</div>
-		</Loading>
+		<UnlockAccountProvider>
```
is this needed for every page in the wallet or just some subset of pages? just thinking on whether or not this is the best place to put this provider haha

pchrysochoidis replied:
Probably most of the pages where we need this use the Layout component but I think moving it to app/global level would make sense as well

williamrobertson13 commented on apps/wallet/src/ui/app/components/accounts/OnboardingFormContext.tsx:
```diff
@@ -0,0 +1,37 @@
+
+const OnboardingFormContext = createContext<OnboardingFormContextType | null>(null);
+
+export const OnboardingFormProvider = ({ children }: { children: ReactNode }) => {
```
this seems pretty reasonable to share the state between forms 👍🏼

williamrobertson13 commented on apps/wallet/src/ui/app/pages/accounts/ImportAccountsPage.tsx:
```diff
@@ -0,0 +1,8 @@
+
+import { Outlet } from 'react-router-dom';
+
+export function ImportAccountsPage() {
```
whut's the point of this?

mamos-mysten replied:
will clean this up!

williamrobertson13 reviewed (approved):
had some small comments here and there but the overall approach looks good 😄

mamos-mysten commented on apps/wallet/src/ui/app/components/accounts/AccountListItem.tsx:
```diff
@@ -36,12 +39,12 @@ export function AccountListItem({
+							isLoading={isLockingOrUnlocking}
+							isLocked={account?.isLocked ?? true}
+							onClick={(e) => {
+								e.stopPropagation();
```
that is correct :) 

I'll make sure I add a comment so that it's clear

mamos-mysten commented on apps/wallet/src/background/account-sources/AccountSource.ts:
```diff
@@ -9,7 +9,7 @@ import {
 import { type Serializable } from '_src/shared/cryptography/keystore';
 
-export type AccountSourceType = 'mnemonic' | 'qredo';
+export type AccountSourceType = 'mnemonic' | 'qredo' | 'zk';
```
this is incorrect

pchrysochoidis commented on apps/wallet/src/ui/app/pages/accounts/ImportPassphrasePage.tsx:
```diff
@@ -28,15 +31,8 @@ export function ImportPassphrasePage() {
-							formValues,
-						);
 						navigate('/accounts/protect-account');
+						setValues({ ...formValues });
```
not sure if it actually matters but I would move this above `navigate`

mamos-mysten commented on apps/wallet/src/ui/app/components/accounts/AccountListItem.tsx:
```diff
@@ -4,16 +4,18 @@
+import { useAccounts } from '../../hooks/useAccounts';
 
 type AccountListItemProps = {
 	address: string;
```
ended up updating this to take in `account`, since we were looking up the account anyway from the `useAccounts` list in order to get props like `isLocked` `nickname` etc.

mamos-mysten commented on apps/wallet/src/ui/app/components/accounts/EditNickname.tsx:
```diff
@@ -19,7 +19,7 @@ import {
 
 const formSchema = z.object({
-	nickname: z.string().trim().nonempty('Required'),
+	nickname: z.string().trim(),
```
this allows you to clear the nickname by saving an empty value

mamos-mysten commented on apps/wallet/src/ui/app/components/accounts/UnlockAccountContext.tsx:
```diff
@@ -0,0 +1,96 @@
+	hideUnlockModal: () => void;
+}
+
+const UnlockAccountContext = createContext<UnlockAccountContextType | undefined>(undefined);
```
indeed!

mamos-mysten commented:
Alright, I addressed most of the PR feedback... I submitted my previous PR a bit to early.... a couple of things to note:

- I changed the default value for `nickname` from `''` to `null` for a couple of reasons: 
    - A default value of `null` is more explicit and in my opinion makes the absence of an account nickname more clear and a bit easier to work with.
    - I tend to use the nullish coalescing operator `??` quite a bit, and with a default value of `''` the  check we're using to derive the account name `nickname ?? domainName ?? formatAddress(account.address)` would evaluate to `nickname` and display an empty string. I feel like this might trip some people up in the future as well.

- Added a basic `<AccountIcon />` component for displaying the account icon and updated the `<SocialGoogle24 />` icon as the current version has a white background baked in.

- In responding to @williamrobertson13's feedback I ended up preferring to pass around `account: SerializedUIAccount` as a prop instead of `accountId` since we end up needing to arrive at that value anyway and ended up doing this in a bunch of places:
```tsx
const accounts = useAccounts()
const account = accounts.find(account => account.id === accountIdFromProps)
```

- Updated the `OnboardingFormContext` name to `AccountsFormContext` since it's also going to be used when adding / importing an account after the wallet is setup as well as during onboarding. Open to suggestions here! 

- Killed the link to Explorer in the `AccountItem` component in favor of copying the address to clipboard, as I found myself accidentally clicking that link constantly which was frustating - we can easily add it back if we need to.

- Allow the `EditNickname` form to accept an empty value, which allows you to "delete" an account nickname by saving an empty value.

There's definitely still some improvements to be made, but this gets things into better shape:

https://github.com/MystenLabs/sui/assets/122397493/3fef7d94-4a80-4ee2-b826-1dc7580f3db1

pchrysochoidis commented on apps/wallet/src/ui/app/components/accounts/AccountListItem.tsx:
```diff
@@ -4,52 +4,47 @@
 					<div className="flex items-center justify-center">
 						<LockUnlockButton
 							isLocked={account.isLocked}
+							isLoading={isLoading}
```
I think it would simplify thing if the `LockUnlockButton` component was using the Unlock context directly instead of accepting props

pchrysochoidis commented on apps/wallet/src/ui/app/pages/site-connect/index.tsx:
```diff
@@ -141,12 +141,7 @@ function SiteConnectPage() {
-											selected={false}
-											address={account.address}
-										/>
+										<AccountListItem key={account.id} selected={false} account={account} />
```
Not for this but at line 134 we should change `accounts` to `unlockedAccounts`

mamos-mysten replied:
updated
